nixos/modules/services/networking/headscale.nix

```diff
@@ -0,0 +1,486 @@
+    db_user = mkNotNull cfg.database.user;
+
+    log_level = cfg.logLevel;
+
+    dns_config = {
+      nameservers = cfg.dns.nameservers;
+      domains = cfg.dns.domains;
+      magic_dns = cfg.dns.magicDns;
+      base_domain = cfg.dns.baseDomain;
+    };
+
+    unix_socket = "${cfg.runDir}/headscale.sock";
+
+    # OpenID Connect
+    oidc = {
+      issuer = cfg.openIdConnect.issuer;
+      client_id = cfg.openIdConnect.clientId;
+      domain_map = cfg.openIdConnect.domainMap;
+    };
+
+    tls_letsencrypt_hostname = mkNotNull cfg.tls.letsencrypt.hostname;
+    tls_letsencrypt_cache_dir = "${cfg.dataDir}/.cache";
+    tls_letsencrypt_challenge_type = mkNotNull cfg.tls.letsencrypt.challengeType;
+    tls_letsencrypt_listen = mkNotNull cfg.tls.letsencrypt.listen;
+    tls_cert_path = mkNotNull cfg.tls.certFile;
+    tls_key_path = mkNotNull cfg.tls.keyFile;
+
+    acl_policy_path = cfg.aclPolicyFile;
+  };
+
+  settingsFormat = pkgs.formats.yaml { };
+  configFile = settingsFormat.generate "headscale.yaml" settings;
+in
+{
+  options = {
+    services.headscale = {
+      enable = mkEnableOption ''
+        Whether to enable the headscale coordination server for Tailscale.
+      '';
+
+      package = mkOption {
+        default = pkgs.headscale;
+        defaultText = literalExpression "pkgs.headscale";
+        type = types.package;
+        description = "headscale package to use.";
+      };
+
+      user = mkOption {
+        default = "headscale";
+        type = types.str;
+        description = ''
+          User the headscale server should execute under.
+        '';
+      };
+
+      group = mkOption {
+        default = "headscale";
+        type = types.str;
+        description = ''
+          If the default user "headscale" is configured then this is the primary
+          group of that user.
+        '';
+      };
+
+      dataDir = mkOption {
+        default = "/var/lib/headscale";
+        type = types.path;
+        description = ''
+          Directory to store data for the headscale server.
+        '';
+      };
+
+      runDir = mkOption {
+        default = "/run/headscale";
+        type = types.path;
+        description = ''
+          Directory for runtime socket.
+        '';
+      };
+
+      serverUrl = mkOption {
+        type = types.str;
+        default = "http://127.0.0.1:8080";
+        description = ''
+          The url clients will connect to.
+        '';
+        example = "https://myheadscale.example.com:443";
+      };
+
+      address = mkOption {
+        type = types.str;
+        default = "127.0.0.1";
+        description = ''
+          Listening address of headscale
+        '';
+        example = "0.0.0.0";
+      };
+
+      port = mkOption {
+        type = types.int;
+        default = 8080;
+        description = ''
+          Listening port of headscale
+        '';
+        example = "443";
+      };
+
+      privateKeyFile = mkOption {
+        default = "";
```

Comment:
is this a mandatory option? if so it's better to not give a default and get an error out of the module system if it's not set